enhancements/network/egress-ip-multi-nic.md

```diff
@@ -0,0 +1,287 @@
+type EgressIPStatusItem struct {
+// Assigned node name
+Node string `json:"node"`
+// Assigned egress IP
+EgressIP string `json:"egressIP"`
+// Assigned network
+Network string `json:"network"`
+}
+```
+
+This will aid debugging of egress IP assignments. Nothing must depend on this network value otherwise upgrades may fail.
+
+### Risks and Mitigations
+- OVN kube node must recalculate all the matched pods for a given egress IP. This will require refactoring this logic
+  from ovnkube-controller and may introduce bugs to egress IP itself.
+- Stale EgressIPs assigned to a non-ovn managed interface
+  A component will be created to manage all network interface address operations.
+  This component will independently ensure the state of all network interfaces is what we expect.
+  Egress IP(s) assigned to an interface will contain a well-known label which will allow us to clean up stale egress IP addresses.
+  Therefore, we don't need any internal caches for this.
+- Stale linux routing rules in the routing policy database control
+  A component will be created to manage all rule operations.
+  We can ensure stale routes are cleaned because we use a specific priority that only we manage, therefore we don't
+  need any internal caches for this.
+  This component will independently manage the operations of rules and ensure they match the state we want.
+- Stale iptables rules
+  A component will be created to manage all iptable operations.
+  This component will independently manage the operations of iptables and ensure they match the state we want.
+  This component will ensure the rules we expect within a newly created specific chain are what we expect.
+  Therefore, we don't need any internal caches for this because we know we manage all rules within this chain.
+- Incorrect egress IP network interface assignment
+  The host addresses label will determine which network (interface) the egress IP will be assigned to.
+  Users may not expect that we will assign this egress IP to whichever interface contains the longest prefix match [1].
+  This is not a problem because we will emit the packet to the correct network regardless.
+- Unable to determine if an egress IP belongs to an ovn managed network or non ovn manged network
+  OVN managed network can be determined from well-known label `k8s.ovn.org/node-primary-ifaddr`
+  Therefore anything outside this subnet mask is consider a "non-ovn" network
+- What about secondary bridges managed by OVN aka br-ex1
+  We ignore them and do not support egress IP.
+- Selecting the wrong network types from interfaces
+  Only consider addresses with scope `Global`.
+
+[1] https://en.wikipedia.org/wiki/Longest_prefix_match
+
+## Design Details
+### Cluster Manager
+If egress IP is within a subnet of a non-ovn managed network, it will need to assign this egress IP to a node which hosts this network.
+
+### OVN kube controller (aka ovn kube master, aka network controller manager)
```

Comment:
So iirc egress IP with shared and lgw both works in a shared gw type of way today. In other words, OVN SNAT is handling the egress IP and it never goes to the kernel. Will you be changing that behavior so that it always goes to the kernel, or is it conditional on if the egress interface is attached to br-ex or not?

If there is some change in behavior, we need to consider how we will upgrade and remove stale things from the GR and/or LRPs from ovn_cluster_router.

Reply:
If the egress IP is still selected to go out br-ex, then it will perform like today and egress on br-ex.

Reply:
but youre right - what could happen is egress IP pre upgrade was selected to go out br-ex, but after upgrade CM found a "better" network to host it using longest prefix matching so in that scenario we would have to clean up LRPs. I added a test case to the enhancement.

Reply:
We could add a caveat to CM selection in that if the EIP falls within the ovn network, then dont use longest prefix match, just select that ovn network... WDYT? That would help us not have to make api changes.

Reply:
Seems like a sensible change to me..

Reply:
Well I'm also asking because im wondering if now is a good time to make egress IP work the right way with local gateway mode. Since you will be basically adding the same logic anyway for multiple nics, we might as well convert local gw to always send to kernel, as it originally should have been.

Reply:
yeah - it maybe a good time to do this but I am not committing to it. We can see in the implementation.